scio-extra/src/main/scala/com/spotify/scio/extra/csv/CsvIO.scala

```diff
@@ -0,0 +1,231 @@
+ * }}}
+ *
+ * =Writing=
+ * == with a header ==
+ * [[https://nrinaudo.github.io/kantan.csv/headers.html Kantan docs]]
+ * {{{
+ *   import kantan.csv._
+ *   import com.spotify.scio.extra.csv._
+ *   implicit val encoder = HeaderEncoder.caseEncoder("fullName", "age")(User.unapply)
+ *
+ *   val users: SCollection[User] = ???
+ *   users.saveAsCsvFile(path)
+ * }}}
+ * == without a header ==
+ * [[https://nrinaudo.github.io/kantan.csv/headers.html Kantan docs]]
+ * {{{
+ *   import kantan.csv._
+ *   import com.spotify.scio.extra.csv._
+ *   implicit val encoder = HeaderEncoder.caseEncoder("fullName", "age")(User.unapply)
+ *
+ *   val users: SCollection[User] = ???
+ *   users.saveAsCsvFile(path)
+ * }}}
+ *
+ */
+object CsvIO {
+
+  implicit val rfc: Coder[ReadableFile] = Coder.beam(new ReadableFileCoder())
+  implicit val metaCoder: Coder[MatchResult.Metadata] = Coder.beam(MetadataCoderV2.of())
+
+  final val DEFAULT_CSV_CONFIG: CsvConfiguration = CsvConfiguration(
+    cellSeparator = ',',
+    quote = '"',
+    quotePolicy = QuotePolicy.WhenNeeded,
+    header = Header.Implicit
+  )
+  final val DEFAULT_READ_PARAMS = CsvIO.ReadParam(compression = beam.Compression.AUTO)
+  final val DEFAULT_WRITE_PARAMS = CsvIO.WriteParam(compression = beam.Compression.UNCOMPRESSED)
+
+  final case class ReadParam(
+    compression: beam.Compression = beam.Compression.AUTO,
+    csvConfiguration: CsvConfiguration = DEFAULT_CSV_CONFIG
+  )
+
+  final case class WriteParam(
+    compression: beam.Compression = beam.Compression.UNCOMPRESSED,
+    csvConfiguration: CsvConfiguration = DEFAULT_CSV_CONFIG,
+    suffix: String = ".csv",
+    numShards: Int = 1
+  ) {
+    def toReadParam: ReadParam =
+      ReadParam(compression = compression, csvConfiguration = csvConfiguration)
+  }
+
+  final case class Read[T: HeaderDecoder: Coder](path: String) extends ScioIO[T] {
+    override type ReadP = CsvIO.ReadParam
+    override type WriteP = CsvIO.WriteParam
+    final override val tapT = TapOf[T]
+
+    override protected def read(
+      sc: ScioContext,
+      params: ReadParam = CsvIO.DEFAULT_READ_PARAMS
+    ): SCollection[T] = CsvIO.read(sc, path, params)
+
+    override protected def write(data: SCollection[T], params: WriteP): Tap[T] =
+      throw new UnsupportedOperationException("Use CsvIO.Write() for writing")
+
+    override def tap(params: ReadP): Tap[T] = new CsvTap[T](path, params)
+  }
+
+  final case class Write[T: HeaderEncoder: Coder](path: String) extends ScioIO[T] {
+    override type ReadP = Nothing // WriteOnly
+    override type WriteP = CsvIO.WriteParam
+    final override val tapT = EmptyTapOf[T]
+
+    override protected def read(sc: ScioContext, params: ReadP): SCollection[T] =
+      throw new UnsupportedOperationException("Use CsvIO.Read() for reading")
+
+    override protected def write(data: SCollection[T], params: WriteP): Tap[Nothing] = {
+      data.applyInternal(csvOut(path, params))
+      EmptyTap
```

Comment:
we can call `tap` here

Reply:
we could but that forces a translation from writeP to readP. nothing complicated but fairly useless

Reply:
tap accepts 1 param which is defined as "Nothing" making it impossible to call.